tools/ClassGen/NodeGen.cpp

```diff
@@ -125,6 +125,8 @@ int main(int argc, char **argv) {
       .addMember(MemberType::Float, "Epsilon")
       .addMember(MemberType::Float, "Momentum")
       .addResult("Input.getType()")
+      .addOverwrittenGradientInput("Mean")
+      .addOverwrittenGradientInput("Var")
```

Comment:
This is only true for when we compile for training, thus I am not sure we want to put that unconditionally.
Put differently, although that's correct, I wonder if we are not leaving performance on the table for inference, which is what we care the most.

What do you think?

Reply:
This is `addOverwrittenGradientInput()`, not `addOverwrittenInput()`, so it does not impact the non-gradient node `BatchNormalizationNode`. It only marks `Mean` and `Var` as overwritten  in `BatchNormalizationGradientNode`. So it should not change anything/negatively impact performance for inference.

Reply:
Ah right, but that's not accurate, isn't it?

I mean those inputs are going to be modified by the lowering of BN when we are in training mode not just by the lowering of BNG. (Admittedly, having one without the other would be pretty strange.)